### PR TITLE
pkg/openthread: update for thread release 20180926

### PR DESCRIPTION
**Pull request by Hyungsin**
### Contribution description
This PR brings OpenThread version released on 20180926 to RIOT-OS.
1. Miscellaneous fixes to match formats.
2. Add MicroTimer function for the future use (CSMA/CA at the OpenThread code).
3. Support data polling for sleepy end devices.
@jia200x / @biboc, please review and test this.

### Testing procedure
I checked that two samr21-xpro boards can be connected and ping to each other.

### CODE REVIEW

**Comment by jia200x** [comment]
awesome!

**Comment by miri64** [comment]
@Hyungsin needs rebase. @jia200x can you give this another look?

**Comment by jia200x** [comment]
@Hyungsin my biggest apologizes. 

I just noticed I personally merged https://github.com/RIOT-OS/RIOT/pull/11540 which also updates OpenThread to the same version.

However, this PR still covers some points that were not included in #11540 (e.g ACK packet handling)
Do you plan to continue working on this? If so, does it work for you if we add the implementation details of this PR in the current master?

Again, sorry for not checking the version :/

**Comment by Hyungsin** [comment]
@jia200x, @miri64, it is frustrating to rebase this PR because another PR, which handles the same issue but posted after this, was reviewed and merged first. This PR got zero reviews meanwhile. 

@jia200x, I can rebase this but it would be good if you can test. Does it work for you?

**Comment by miri64** [comment]
> @jia200x, @miri64, it is frustrating to rebase this PR because another PR #11540, which handles the same issue but posted after this, was reviewed and merged first. This PR got zero reviews meanwhile.

Sorry, I wasn't aware about this happening. We are aware of this problem and try to be more careful in the future (see https://github.com/RIOT-OS/RIOT/pull/12543). I hand the review to @jia200x.

**Comment by Hyungsin** [comment]
@jia200x @miri64 I rebased it. Hope that it is tested soon.

File: pkg/openthread/contrib/platform_alarm.c [hunk 1]
```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
...
+}
+
+/**
+ * Set the alarm to fire at @p aDt microseconds after @p aT0.
+ *
+ * @param[in] aInstance  The OpenThread instance structure.
+ * @param[in] aT0        The reference time.
+ * @param[in] aDt        The time delay in microseconds from @p aT0.
+ */
+void otPlatAlarmMicroStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
```
Comment (maribu) [inline pkg/openthread/contrib/platform_alarm.c]: I have never worked with openthread so far, so please forgive if I'm asking nonsense.

The implementation in this PR would allow one timer in total to be set with `otPlatAlamMicroStartAt()` and one with the milliseconds version. (And the version in master would allow one in total.) This seems to be strange to me.

If I had to guess, I would say one timer per `otInstance` would make most sense. And one timer in total would make second-most sense to. Can someone with the required knowledge comment on this?

I think it should be possible to only provide one common internal implementation and offer `otPlatAlarmMicroStartAt()` and `otPlatAlarmMilliStartAt()` as `static inline` wrappers on top if this in order to safe some ROM.

File: pkg/openthread/contrib/platform_alarm.c [hunk 2]
```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
...
+{
+    (void)aInstance;
+    DEBUG("openthread: otPlatAlarmMicroStartAt: aT0: %" PRIu32 ", aDT: %" PRIu32 "\n", aT0, aDt);
+    xtimer_set(&ot_microtimer, aDt);
+}
+
+/* OpenThread will call this to stop alarms */
+void otPlatAlarmMicroStop(otInstance *aInstance)
```
Comment (maribu) [inline pkg/openthread/contrib/platform_alarm.c]: This should also be possible to combine with the milli version (e.g. by one being a `static inline` wrapper for the other). Of course only if not two timers in total is actually what openthread expects.

File: pkg/openthread/contrib/platform_alarm.c [hunk 3]
```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
...
+{
+    (void)aInstance;
+    DEBUG("openthread: otPlatAlarmMicroStop\n");
+    xtimer_remove(&ot_microtimer);
+}
+
+/* OpenThread will call this for getting running time in microsecs */
+uint32_t otPlatAlarmMicroGetNow(void)
```
Comment (maribu) [inline pkg/openthread/contrib/platform_alarm.c]: Providing this as `static inline` wrapper should safe some ROM.

File: pkg/openthread/contrib/platform_radio.c [hunk 4]
```diff
@@ -35,11 +36,17 @@
 #include "debug.h"
 
 #define RADIO_IEEE802154_FCS_LEN    (2U)
+#define IEEE802154_ACK_LENGTH       (5U)
+#define IEEE802154_DSN_OFFSET       (2U)
 
 static otRadioFrame sTransmitFrame;
 static otRadioFrame sReceiveFrame;
+static otRadioFrame sAckFrame;
```
Comment (maribu) [inline pkg/openthread/contrib/platform_radio.c]: Again this is in misalignment with RIOT's coding convention. (As are the one above, that are not related to this PR.) If there is not technical reason for this, this should be renamed in order to comply with the coding convention.

File: pkg/openthread/contrib/platform_radio.c [hunk 5]
```diff
@@ -172,20 +182,35 @@ void recv_pkt(otInstance *aInstance, netdev_t *dev)
...
+    }
+    sAckFrame.mPsdu[1] = 0;
+    sAckFrame.mPsdu[2] = sTransmitFrame.mPsdu[IEEE802154_DSN_OFFSET];
+
+    sAckFrame.mInfo.mRxInfo.mRssi = OT_RADIO_RSSI_INVALID;
+}
+
 /* Called upon TX event */
-void send_pkt(otInstance *aInstance, netdev_t *dev, netdev_event_t event)
+void sent_pkt(otInstance *aInstance, netdev_t *dev, netdev_event_t event)
```
Comment (maribu) [inline pkg/openthread/contrib/platform_radio.c]: ~~Why this change in name?~~
Please see comment in declaration.

File: pkg/openthread/contrib/platform_radio.c [hunk 6]
```diff
@@ -172,20 +182,35 @@ void recv_pkt(otInstance *aInstance, netdev_t *dev)
...
         case NETDEV_EVENT_TX_COMPLETE:
             DEBUG("openthread: NETDEV_EVENT_TX_COMPLETE\n");
-            otPlatRadioTxDone(aInstance, &sTransmitFrame, NULL, OT_ERROR_NONE);
+            _create_fake_ack_frame(false);
+            otPlatRadioTxDone(aInstance, &sTransmitFrame, &sAckFrame, OT_ERROR_NONE);
             break;
         case NETDEV_EVENT_TX_COMPLETE_DATA_PENDING:
             DEBUG("openthread: NETDEV_EVENT_TX_COMPLETE_DATA_PENDING\n");
-            otPlatRadioTxDone(aInstance, &sTransmitFrame, NULL, OT_ERROR_NONE);
+            _create_fake_ack_frame(true);
```
Comment (maribu) [inline pkg/openthread/contrib/platform_radio.c]: This seems odd to me. Maybe it is only because of the naming and me not being familiar with openthread. But I would expect ACKs to be created upon receiving a frame, not after sending one.

File: pkg/openthread/contrib/platform_radio.c [hunk 7]
```diff
@@ -213,17 +238,17 @@ void otPlatRadioSetExtendedAddress(otInstance *aInstance, const otExtAddress *aE
 {
     (void)aInstance;
     DEBUG("openthread: otPlatRadioSetExtendedAddress\n");
-    char reversed_addr[IEEE802154_LONG_ADDRESS_LEN];
+    uint8_t reversed_addr[IEEE802154_LONG_ADDRESS_LEN];
     for (unsigned i = 0; i < IEEE802154_LONG_ADDRESS_LEN; i++) {
-        reversed_addr[i] = (uint8_t) ((uint8_t *)aExtAddress)[IEEE802154_LONG_ADDRESS_LEN - 1 - i];
+        reversed_addr[i] = aExtAddress->m8[IEEE802154_LONG_ADDRESS_LEN - 1 - i];
```
Comment (maribu) [inline pkg/openthread/contrib/platform_radio.c]: This is network vs host byte order, right? If so, `reversed_addr` is a misleading name. (You could also take a look at the `eui64_t` type and see if it would be handy here.) 

If the host byte order happens to be network byte order, this would not work. Maybe `_set_long_addr()` could be changed to always expect network byte order? (Or even `eui64_t`?)

File: pkg/openthread/include/ot.h [hunk 8]
```diff
@@ -110,7 +113,7 @@ void recv_pkt(otInstance *aInstance, netdev_t *dev);
  * @param[in]  dev                pointer to a netdev interface
  * @param[in]  event              just occurred netdev event
  */
-void send_pkt(otInstance *aInstance, netdev_t *dev, netdev_event_t event);
+void sent_pkt(otInstance *aInstance, netdev_t *dev, netdev_event_t event);
```
Comment (maribu) [inline pkg/openthread/include/ot.h]: OK. Now (in context of the API doc I get the rename). Maybe even changing it to `pkt_sent()`?

File: pkg/openthread/include/ot.h [hunk 9]
```diff
@@ -123,8 +126,9 @@ void openthread_bootstrap(void);
  * @param[in]  dev                pointer to a netdev interface
  * @param[in]  tb                 pointer to the TX buffer designed for OpenThread
  * @param[in]  rb                 pointer to the RX buffer designed for Open_Thread
+ * @param[in]  ab                 pointer to the ACK buffer designed for Open_Thread
```
Comment (maribu) [inline pkg/openthread/include/ot.h]: Shouldn't it be "designated" instead of "designed"? If so, would you mind to sneak in a fix of the doc of `tb` and `rb` (as separate commit) into this PR?

File: pkg/openthread/contrib/platform_radio.c (hunk 6 above)
Comment (Hyungsin) [reply pkg/openthread/contrib/platform_radio.c]: Yes, this "fake" ack is created to report the result of a sent frame to the upper layer. This is not for responding to a received frame.

File: pkg/openthread/include/ot.h (hunk 8 above)
Comment (Hyungsin) [reply pkg/openthread/include/ot.h]: Yes. This is called after a packet transmission finishes. So may be pkt_sent is more appropriate.

File: pkg/openthread/include/ot.h (hunk 9 above)
Comment (Hyungsin) [reply pkg/openthread/include/ot.h]: Either one makes sense to me. So I can change as you want.

File: pkg/openthread/contrib/platform_radio.c (hunk 5 above)
Comment (Hyungsin) [reply pkg/openthread/contrib/platform_radio.c]: Checked

File: pkg/openthread/contrib/platform_alarm.c (hunk 1 above)
Comment (Hyungsin) [reply pkg/openthread/contrib/platform_alarm.c]: We need to call otPlatAlarmMilliFired(sInstance) when a millitimer is fired and otPlatAlarmMicroFired(sInstance) when a microtimer is fired. Without separation, I think it is hard to know what timer is fired. FYI, https://github.com/openthread/openthread/blob/master/src/core/common/timer.cpp

**Comment by jia200x** [comment]
> @jia200x, @miri64, it is frustrating to rebase this PR because another PR #11540, which handles the same (actually smaller) issue but posted 2 months after this, was reviewed and merged first. Surprisingly, it took only two days for the PR #11540 to be merged, but this PR got zero review for the last 8 months.

Hi @Hyungsin.
I totally get your frustration. My apologies, I was the one who merged #11540. There were not bad intentions, it was my mistake for not (re)checking existing solutions without approving and merging.

As @miri64 pointed out, we are aware of this situation and trying to fix it asap.

Thanks anyway that besides this, you rebased it to the current master. I will run the tests ASAP and we would be ready to go.

**Comment by Hyungsin** [comment]
> I will run the tests ASAP and we would be ready to go.

It is OK and Thanks for testing!

**Comment by Hyungsin** [comment]
@jia200x ping?

**Comment by jia200x** [comment]
> @jia200x ping?

I will now. I was busy with #11022

File: pkg/openthread/contrib/netdev/openthread_netdev.c [hunk 10]
```diff
@@ -39,6 +42,7 @@ static msg_t _queue[OPENTHREAD_QUEUE_LEN];
 
 static kernel_pid_t _pid;
 static otInstance *sInstance;
+static bool otTaskPending = false;
```
[the remark this answers is not in the document]
Comment (jia200x) [reply pkg/openthread/contrib/netdev/openthread_netdev.c]: we should update this in the file as well.
I would be ok to leave it as is in this PR and immediately after open another one for fixing this. What do you think?

File: pkg/openthread/contrib/platform_radio.c (hunk 6 above)
Comment (jia200x) [reply pkg/openthread/contrib/platform_radio.c]: This same technique is used in some of the official platform implementations of Openthread.

File: pkg/openthread/contrib/platform_alarm.c [hunk 11]
```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
...
 {
-    uint32_t now = xtimer_now_usec() / US_PER_MS;
-    DEBUG("openthread: otPlatAlarmGetNow: %" PRIu32 "\n", now);
+    uint32_t now = (uint32_t)(xtimer_now_usec64() / US_PER_MS);
+    DEBUG("openthread: otPlatAlarmMilliGetNow: %" PRIu32 "\n", now);
+    return now;
+}
+
+/* Interupt handler for OpenThread micro-timer event */
+static void _microtimer_cb(void* arg) {
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply pkg/openthread/contrib/platform_alarm.c]: resolved!

File: pkg/openthread/contrib/platform_alarm.c [hunk 12]
```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
...
-
-    if (aDt == 0) {
-        msg_send(&ot_alarm_msg, thread_getpid());
-    }
-    else {
-        int dt = aDt * US_PER_MS;
-        xtimer_set_msg(&ot_timer, dt, &ot_alarm_msg, thread_getpid());
-    }
+    uint32_t dt = aDt * US_PER_MS;
+    xtimer_set(&ot_millitimer, dt);
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply pkg/openthread/contrib/platform_alarm.c]: resolved!

**Comment by Hyungsin** [comment]
@jia200x just resolved your comments. Thanks!

**Comment by aabadie** [comment]
@jia200x, maybe you'll find some time to have another look here ?

**Comment by jia200x** [comment]
I'm back into this, I'm running some tests now.

File: pkg/openthread/contrib/platform_radio.c (hunk 7 above)
Comment (jia200x) [reply pkg/openthread/contrib/platform_radio.c]: OpenThread uses IEEE addresses in little endian (see [OpenThread radio API](https://github.com/openthread/openthread/blob/295ce1125ec6b0b38a2de90d0a6e54d010a5a061/include/openthread/platform/radio.h#L341))

So we need to reverse in order to write it in the hardware. Maybe we could choose a better name for this variable or add comments?

File: pkg/openthread/contrib/platform_radio.c (hunk 4 above)
Comment (jia200x) [reply pkg/openthread/contrib/platform_radio.c]: We have to adapt this file anyway in a further PR (lots of Camel Case here). Is it OK for you to leave it as is and we open a follow up right after we merge this?

File: pkg/openthread/contrib/platform_alarm.c (hunk 3 above)
Comment (jia200x) [reply pkg/openthread/contrib/platform_alarm.c]: This is called by OpenThread, the symbol has to be exposed. So I guess it couldn't be `static inline` :/

**Comment by miri64** [comment]
@Hyungsin @jia200x any progress?

**Comment by jia200x** [comment]
ping @Hyungsin

**Comment by benpicco** [comment]
Should probably be updated to `thread-reference-20191113` and the new Radio HAL.

**Comment by jia200x** [comment]
> Should probably be updated to thread-reference-20191113 and the new Radio HAL.

Ideally. 
OpenThread can run on top of the SubMAC (see https://github.com/RIOT-OS/RIOT/pull/14979), so then we don't have to take care of some "high level" stuff (getting ACK frames and handing it in to the stack, etc).
